### PR TITLE
Refactor Desul atomics to support compiling CUDA with nvc++

### PR DESCRIPTION
**Pull request by dalg24**
Not asking for feedback at this time

### CODE REVIEW

**Comment by dalg24** [comment]
Retest this please

**Comment by dalg24** [comment]
Retest this please

**Comment by dalg24** [comment]
Retest this please

**Review by crtrott — APPROVED** [review]

File: tpls/desul/include/desul/atomics/Generic.hpp [hunk 1]
```diff
@@ -16,10 +16,11 @@ SPDX-License-Identifier: (BSD-3-Clause)
 
 #include <type_traits>
 
+#include <desul/atomics/Fetch_Op.hpp>
+
 #include <desul/atomics/Generic_Device.hpp>
 #include <desul/atomics/Generic_Host.hpp>
 
-#include <desul/atomics/Fetch_Op.hpp>
```
Comment (dalg24) [inline tpls/desul/include/desul/atomics/Generic.hpp]: @crtrott this is actually awkward.  The latest version of the PR on the desul side had these include reordered with Clang-Format and the alphabetic order happens to do what we want...

**Comment by dalg24** [comment]
This is ready

**Comment by masterleinad** [comment]
Did you do a plain copy from `desul`?

**Comment by dalg24** [comment]
`desul/atomics/Lock_Array_Cuda.hpp -> desul/atomics/Lock_Array_CUDA.hpp`

**Comment by dalg24** [comment]
I withdrew the change to the CI.  Will tackle that in a follow up PR.

**Comment by dalg24** [comment]
> Did you do a plain copy from `desul`?

I cleaned up history.
* The first commit is a copy from desul [f5b9eeb](https://github.com/desul/desul/commit/f5b9eeb740ed855ff47ba006f27921734a382737), 
* the second one a bug fix corresponding to desul/desul#71, 
* and the third one a minor update on the Kokkos side

File: core/src/Cuda/Kokkos_Cuda_Locks.hpp [hunk 2]
```diff
@@ -54,7 +54,7 @@
 #include <Cuda/Kokkos_Cuda_Error.hpp>
 
 #ifdef KOKKOS_ENABLE_IMPL_DESUL_ATOMICS
-#include <desul/atomics/Lock_Array_Cuda.hpp>
+#include <desul/atomics/Lock_Array_CUDA.hpp>
```
Comment (Rombur) [inline core/src/Cuda/Kokkos_Cuda_Locks.hpp:57]: Why is the name different? In master desul the name is still `Lock_Array_Cuda.hpp`: https://github.com/desul/desul/blob/master/atomics/include/desul/atomics/Lock_Array_Cuda.hpp

Comment (dalg24) [reply core/src/Cuda/Kokkos_Cuda_Locks.hpp:57]: You are right.  I will fix on the desul side.  Thank you.

**Review by Rombur — APPROVED** [review]

File: core/src/Cuda/Kokkos_Cuda_Locks.hpp (hunk 2 above)
Comment (dalg24) [reply core/src/Cuda/Kokkos_Cuda_Locks.hpp:57]: Updated

**Pull request merged**